### PR TITLE
chore(build): add build scripts to build all or individual projects

### PR DESCRIPTION
**Pull request by jeffbcross**
Paired with @gdi2290 to clean up build. Still some cleanup to to on separate modules, hope to wrap that up this evening.

TODO
- [x] Prune npm packages
- [ ] Move changelog generation to gulpfile
- [ ] Update old build system readme to reflect changes
- [ ] Publish all modules with `@next` tag and get folks to dogfood.
- [x] Remove examples folder (and open issue to build fresh examples)
- [x] Add unit tests to travis build
- [x] Add script to publish all modules to npm (with optional tag)
- [x] Make sure d.ts files are compatible with latest stable version of TypeScript.
- [x] Make individual package.jsons dynamic to use versions from root package.json
- [x] move preboot to separate repo: angular/preboot
- [x] update preboot links to point to new repo
- [x] Remove tsconfig from each individual module to rely on global config.
- [x] Clean up .gitignore to not allow `dist`, `node_modules`, or `typings` inside of individual folders
- [x] Rename `src` to `dist` in each built module (or just flatten files to be in root of package)
- [x] Add script to copy package.jsons to outputs for publishing
- [x] Add an option to build all tests along with modules (anything matching `.spec.ts`)
- [x] Add script to copy built files to `tasks` inside `grunt-prerender`
 
I propose moving these tasks outside of this PR since it's already quite substantial (and so @gdi2290 can do it instead of me :))
- [ ] Make `npm start` start up example app
- [ ] Clean up unnecessary files in root directory
- [ ] Remove all gulp artifacts (except for what's in modules/preboot, and packages that are required by preboot).

### CODE REVIEW

**Comment by MarkPieszak** [comment]
I know I have all those individual scripts in there for each module sorry (messy I know), it was just to get the install:all script working for windows/Linux, the --ignore part is needed just to make npm not stop its lifecycle if tsc has any minor error. Just a fyi!

**Comment by jeffbcross** [comment]
Pushed some more progress, still a lot of work!

File: build-utils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  'angular2-gulp-prerender': '~1.0.0',
+  'angular2-hapi-engine': '~1.0.0',
+  'angular2-universal-polyfills': '~1.0.0',
+  'angular2-universal': '~1.0.0',
+  'angular2-webpack-prerender': '~1.0.0'
+};
+
+const config = require('./tsconfig.json');
+
+xdescribe('build-utils', () => {
```
Comment (gdi2290) [inline build-utils.spec.ts]: we can remove the `x` now right

File: build-utils.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+export function getCompilerOptions(config: TSConfig): ts.CompilerOptions  {
+  return ts.convertCompilerOptionsFromJson(config.compilerOptions, __dirname, 'tsconfig.json').options;
+}
+
+export function getPublishedModuleNames(allModules: string[]): string[] {
+  return allModules.map(name => {
+    if (name === 'polyfills') {
+      return 'angular2-universal-polyfills';
```
Comment (gdi2290) [inline build-utils.ts]: we should probably rename the folder to `universal-polyfills` to keep it consistent

File: build-utils.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+const glob = require('glob');
+const rimraf = require('rimraf');
+
+export function removeDist(compilerOptions) {
+  rimraf.sync(compilerOptions.outDir);
+  assert(!fs.existsSync('./dist'), 'dist directory was NOT removed');
+}
+
+export function getAllModules(): string[] {
+  return fs.readdirSync('./modules');
```
Comment (gdi2290) [inline build-utils.ts]: we should probably create a constant for the modules path for example if we change the folder name to `src/` later

`const MODULES_PATH = './modules';`

File: build-utils.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
+import * as fs from 'fs';
+import { execSync } from 'child_process';
+import { relative, resolve} from 'path';
```
Comment (gdi2290) [inline build-utils.ts]: can we keep this as path.resolve or path.relative if we need the namespace we can say `* as nodePath`

File: build.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+/// <reference path="./typings/index.d.ts" />
```
Comment (gdi2290) [inline build.ts]: we don't need this file anymore because of our new gulpfile correct?

File: gulpfile.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      .pipe(rename(stripSrcFromPath))
+      .pipe(gulp.dest('dist'))
+      .on('end', maybeDone),
+    output.dts
+      .pipe(rename(stripSrcFromPath))
+      .pipe(gulp.dest('dist'))
+      .on('end', maybeDone)
+
+    function maybeDone() {
+      if (++doneCount) res();
```
Comment (gdi2290) [inline gulpfile.ts]: we can change this to make it more clear of our intent?
```es6
doneCount += 1;
if (doneCount >= 2) {
  return res(); // resolve
}
```

File: gulpfile.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+function build(path: string[]): Promise<any> {
+  // TODO: remove this if it every works in watch mode without needing to re-create.
+  let project = gulpTs.createProject('tsconfig.json', {
+    typescript: require('typescript'),
+    rootDir: 'modules'
+  });
+  let output = gulp.src(path)
+    .pipe(gulpTs(project));
+  return new Promise((res, rej) => {
```
Comment (gdi2290) [inline gulpfile.ts]: `res` = > `resolve`

File: gulpfile.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+          data = Object.assign({}, data, {
+            [k]: buildUtils.getRootDependencyVersion(Object.keys(data[k]), rootDependencies)
+          })
+        });
+      return buildUtils.addMetadataToPackage(data, rootPkg);
+    }, 2))
+    .pipe(gulp.dest('dist'));
+});
+
+function stripSrcFromPath(path) {
```
Comment (gdi2290) [inline gulpfile.ts]: we should move this to buildUtils

File: modules/universal/src/browser/env.ts [hunk 9]
```diff
@@ -1,2 +1,4 @@
 export const isBrowser = true;
 export const isNode = false;
+
+console.log('hi');
```
Comment (gdi2290) [inline modules/universal/src/browser/env.ts]: oops we forgot to remove this

File: tslint.json [hunk 10]
```diff
@@ -1,54 +0,0 @@
-{
```
Comment (gdi2290) [inline tslint.json]: no tslint.json ?

File: build-utils.ts (hunk 3 above)
Comment (jeffbcross) [inline build-utils.ts]: Will do the same for `dist/` and the `angular2-` prefix.

File: modules/universal/src/browser/env.ts (hunk 9 above)
Comment (jeffbcross) [inline modules/universal/src/browser/env.ts]: I forgot nothing.

File: tslint.json (hunk 10 above)
Comment (jeffbcross) [inline tslint.json]: Restored it.

**Comment by gdi2290** [comment]
LGTM

**Pull request merged**